mitodl/micromasters
# Exams: result/request file audit trail

#### What are the relevant tickets?
Fixes #2896 and fixes #2895 

#### What's this PR do?
This adds "auditing" (encrypting and storing of files to S3) for files we send and receive for exams.

#### How should this be manually tested?

###### Set up your .env
- First, you need to generate a GPG key `gpg --gen-key`. Follow the prompts.
- Get the key fingerprint `gpg --list-keys --fingerprint`, put this in `.env` as `EXAMS_AUDIT_ENCRYPTION_FINGERPRINT`
- Export the gpg key `gpg --armor --export <EMAIL> | base64 -w 0` (use email you entered in step 1), put this as `EXAMS_AUDIT_ENCRYPTION_PUBLIC_KEY` 
- Requisition yourself values for `EXAMS_AUDIT_S3_BUCKET`, `EXAMS_AUDIT_AWS_ACCESS_KEY_ID`, and `EXAMS_AUDIT_AWS_SECRET_ACCESS_KEY` from me

###### Test it in shell
Easiest way to test this is in shell, because otherwise you have to simulate the entire request/response flow locally which is a pain.

- First, write a test file to your project root. Can have whatever contents.
- In `./manage.py shell`:

```python
from exams.pearson.audit import ExamDataAuditor
from django.test import override_settings
auditor.configure() # should return without error
with override_settings(EXAMS_AUDIT_ENABLED=True):
     auditor.audit_request('/src/my_test_file.txt') # replace with your unique filename
```
- At this point, the file should be on the RC S3 bucket. To verify:
```python
b = auditor.store.get_bucket()
keys = b.get_all_keys(max_keys=100)
keys
```
- You should see a key with your file printed under the `exam_audits/request/` path.
- To verify contents:
```python
with open('/src/encrypted_file.gpg', 'wb') as f:
   key = keys[0] # whatever index it appears in
   key.get_contents_to_file(f)
```
- You should now have `encrypted_file.gpg` in your project root. `cat` to verify it isn't plaintext.
- Decrypt it to be sure it was encrypted correctly: `gpg --output decrypted.txt --decrypt encrypted.gpg`
- Clean up your all these files once you've verified, sorry for the mess.

#### Where should the reviewer start?
exams/pearson/audit.py

#### Any background context you want to provide?
This provides a trail of what happened historically in case questions come up regarding what we sent or what we received.

#### What GIF best describes this PR or how it makes you feel?
https://giphy.com/gifs/JTSs32Pa9iXYc/
